discovery_engine_core/bindings/src/types/embedding.rs

```diff
@@ -0,0 +1,119 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+//! FFI functions for handling embeddings.
+
+use core::document::Embedding;
+
+use ndarray::Array;
+
+use crate::types::slice::boxed_slice_from_raw_parts;
+
+/// Creates a rust `Embedding` with given capacity at given memory address.
+///
+/// # Safety
+///
+/// - It must be valid to write an `Embedding` instance to given pointer.
+/// - The passed in slice must represent a `Box<[f32]>` and transfers ownership,
+///   it must be fully initialized.
+#[no_mangle]
+pub unsafe extern "C" fn init_embedding_at(
+    place: *mut Embedding,
+    owning_ptr: *mut f32,
+    len: usize,
+) {
+    let vec = unsafe { boxed_slice_from_raw_parts::<f32>(owning_ptr, len).into() };
+    let embedding = Embedding::from(Array::from_vec(vec));
+    unsafe {
+        place.write(embedding);
+    }
+}
+
+/// Returns a pointer to the begin of the `[f32]` backing the `Embedding`
+///
+/// # Safety
+///
+/// The pointer must point to a valid [`Embedding`] instance.
+#[no_mangle]
+pub unsafe extern "C" fn get_embedding_buffer(embedding: *const Embedding) -> *const f32 {
+    unsafe { &*embedding }.as_ptr()
+}
+
+/// Returns len of the given embedding.
+///
+/// # Safety
+///
+/// The pointer must point to a valid [`Embedding`] instance.
+#[no_mangle]
+pub unsafe extern "C" fn get_embedding_buffer_len(embedding: *mut Embedding) -> usize {
+    //WARNING: This holds for `Embedding` but not for all possible `ndarray::Array<...>`.
```

Comment:
In our case `Embedding` only can have cases where `len` == "buffer len".

But it's a good point that we should handle that more explicitly/put in some fail safe.

(I had the misconception that this constraints is implicitly statically encoded in `Embedding`, and hence this isn't needed, but it is not.)

Reply:
I changed it to:

`unsafe { &*embedding }.as_slice().unwrap_or_else(|| abort()).len()`

with following note:

```rust
/// # Aborts
///
/// Aborts if the embedding is not "contiguous and in standard order".
///
//Note: Wether this is or isn't the case should (for our use case) be always
// the same independent of input data. Hence this should be caught by
// tests. If that isn't the case anymore it should be considered to
// change this interface, e.g. to support reading a buffer with strides.
```

Through this note is based on some older understanding of our algorithms and they have changed since then.
Maybe we should just teach dart to parse strides?